framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java

```diff
@@ -768,6 +760,103 @@ public static AnnotatedTypeMirror leastUpperBound(
+      return glbSubtype(atypeFactory.getQualifierHierarchy(), type1, type2);
+    } else if (types.isSameType(type2.getUnderlyingType(), glbJava)) {
+      return glbSubtype(atypeFactory.getQualifierHierarchy(), type2, type1);
+    }
+
+    if (glbJava.getKind() != TypeKind.INTERSECTION) {
+      // If one type isn't a subtype of the other, then GLB must be an intersection.
+      throw new BugInCF(
+          "AnnotatedTypes#greatestLowerBound: unexpected java type: %s. type1: %s, type2: %s",
+          glbJava, type1, type2);
+    }
+    QualifierHierarchy qualifierHierarchy = atypeFactory.getQualifierHierarchy();
+    Set<AnnotationMirror> setA =
+        AnnotatedTypes.findEffectiveLowerBoundAnnotations(qualifierHierarchy, type1);
+    Set<AnnotationMirror> setB =
+        AnnotatedTypes.findEffectiveLowerBoundAnnotations(qualifierHierarchy, type2);
+    Set<? extends AnnotationMirror> glbAnno = qualifierHierarchy.greatestLowerBounds(setA, setB);
+
+    AnnotatedIntersectionType glb =
+        (AnnotatedIntersectionType) AnnotatedTypeMirror.createType(glbJava, atypeFactory, false);
+
+    for (AnnotatedTypeMirror bound : glb.getBounds()) {
+      if (types.isSameType(bound.getUnderlyingType(), type1.getUnderlyingType())) {
+        atypeFactory.replaceAnnotations(type1, bound);
+      } else if (types.isSameType(bound.getUnderlyingType(), type2.getUnderlyingType())) {
+        atypeFactory.replaceAnnotations(type2, bound);
+      } else {
+        throw new BugInCF(
+            "Neither %s nor %s is one of the intersection bounds in %s. Bound: %s",
+            type1, type2, bound, glb);
+      }
+    }
+
+    glb.addAnnotations(glbAnno);
+    return glb;
+  }
+
+  /**
+   * Returns the annotated greatest lower bound of {@code subtype} and {@code supertype}, where the
+   * underlying java types are in a subtyping relationship.
+   *
+   * @param qualifierHierarchy QualifierHierarchy
+   * @param subtype annotated types whose underlying type is a subtype of {@code supertype}
+   * @param supertype annotated types whose underlying type is a supertype of {@code subtype}
+   * @return the annotated greatest lower bound of {@code subtype} and {@code supertype}
+   */
+  private static AnnotatedTypeMirror glbSubtype(
+      QualifierHierarchy qualifierHierarchy,
+      AnnotatedTypeMirror subtype,
+      AnnotatedTypeMirror supertype) {
+    AnnotatedTypeMirror glb = subtype.deepCopy();
+    glb.clearAnnotations();
+
+    for (AnnotationMirror top : qualifierHierarchy.getTopAnnotations()) {
+      AnnotationMirror subAnno = subtype.getAnnotationInHierarchy(top);
+      AnnotationMirror superAnno = supertype.getAnnotationInHierarchy(top);
+      if (subAnno != null && superAnno != null) {
+        glb.addAnnotation(qualifierHierarchy.greatestLowerBound(subAnno, superAnno));
+      } else if (subAnno == null && superAnno == null) {
+        assert subtype.getKind() == TypeKind.TYPEVAR && supertype.getKind() == TypeKind.TYPEVAR;
```

Comment:
I suggest making these assertions (here and 2 lines down) into `if (...) throw new BugInCF()`, so we get earlier notification of problems.